### PR TITLE
New AS2 Handlers

### PR DESCRIPTION
**Pull request by jadbalout**

### CODE REVIEW

**Comment by widd** [comment]
So far so good - just add a newline between each message/packet ID and the description.

**Review by widd — CHANGES_REQUESTED** [review]
Lots of little things I see right off the bat - will repost with details.

File: as2/server/redemption.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Redemption
```
Comment (widd) [inline as2/server/redemption.md]: Will need a secondary confirmation on this as I am not that familiar with the redemption protocol. Just as well, "unknown & not needed" needs to be 100% accurately confirmed before I can accept this. I have only encountered one instance in the protocol where this was true so far.

File: as2/server/player.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+##### Example
+`%xt%sf%0%123%5%`
+
+### Send Move
+Sent to all penguins in a room to update a player location.
+
+|Argument|Type|Variadic?|
+|---|---|---|
+|penguinID|Unsigned integer|No|
+|X Postion|Unsigned integer|No|
```
Comment (widd) [inline as2/server/player.md]: "Position" - simple spelling error should be fixed.
Suggested name: "positionX".

File: as2/server/player.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+|Y Postion|Unsigned integer|No|
```
Comment (widd) [inline as2/server/player.md]: "Position" - simple spelling error should be fixed.
Suggested name: "positionY".

File: as2/server/player.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+##### Example
+`%xt%sp%0%123%0%0%`
+
+### Send Message
+Sent to all penguins in a room to show a message sent by a penguin.
+
+|Argument|Type|Variadic?|
+|---|---|---|
+|penguinID|Unsigned integer|No|
+|Message|String|No|
```
Comment (widd) [inline as2/server/player.md]: Lowercase argument names for table readability.

File: as2/server/errors.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+---
+### Error
+Sent to a penguin to show a specific error. `e`
+
+|Argument|Type|Variadic?|
+|---|---|---|
+|Error ID|Unsigned integer|No|
+
+##### Error Types
+There is a large amount of error IDs in Club Penguin, here is a list of most(if not all) of them. https://pastebin.com/uGZ2QPAs
```
Comment (widd) [inline as2/server/errors.md]: The pastebin should be removed. The CP Protocol project should be self-contained and entirely accessible offline - we can document "official" error IDs (that the live CP server used) in a section at the end of the `errors` document.

File: as2/server/player.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
+# Player
+##### Messages pertaining to the player actions.
+---
+### Send Frame
+Sent to all penguins in a room to update a player frame. `sf`
+
+|Argument|Type|Variadic?|
+|---|---|---|
+|penguinID|Unsigned integer|No|
+|Frame|Unsigned integer|No|
```
Comment (widd) [inline as2/server/player.md]: Lowercase argument names for table readability (Frame -> frame).

File: as2/server/player.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+##### Example
+`%xt%sp%0%123%0%0%`
+
+### Send Throw Snowball
+Sent to all penguins in a room to show a snowball thrown by a penguin. `sb`
+
+|Argument|Type|Variadic?|
+|---|---|---|
+|penguinID|Unsigned integer|No|
+|X Postion|Unsigned integer|No|
```
Comment (widd) [inline as2/server/player.md]: "Position" - simple spelling error should be fixed.
Suggested name: "positionX".

File: as2/server/player.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+|Y Postion|Unsigned integer|No|
```
Comment (widd) [inline as2/server/player.md]: "Position" - simple spelling error should be fixed.
Suggested name: "positionY".

File: as2/server/player.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+`%xt%sb%0%123%0%0%`
+
+
+### Send Emote
+Sent to all penguins in a room to show an emote sent by a penguin. `se`
+
+|Argument|Type|Variadic?|
+|---|---|---|
+|penguinID|Unsigned integer|No|
+|Emote ID|Unsigned integer|No|
```
Comment (widd) [inline as2/server/player.md]: Per previous review comments:
Emote ID -> emoteID
Consistency!

File: as2/server/player.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+##### Example
+`%xt%se%0%123%3%`
+
+### Send Joke
+Sent to all penguins in a room to show a joke sent by a penguin. `sj`
```
Comment (widd) [inline as2/server/player.md]: "joke sent" should be "joke said" to better describe the client's visual behaviour.

File: as2/server/player.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+|Argument|Type|Variadic?|
+|---|---|---|
+|penguinID|Unsigned integer|No|
+|Joke ID|Unsigned integer|No|
```
Comment (widd) [inline as2/server/player.md]: Joke ID -> jokeID

File: as2/server/player.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+`%xt%sj%0%123%3%`
+
+
+### Send Message
+Sent to all penguins in a room to show a message sent by a penguin. `sm`
+
+|Argument|Type|Variadic?|
+|---|---|---|
+|penguinID|Unsigned integer|No|
+|Message|String|No|
```
Comment (widd) [inline as2/server/player.md]: Message -> message

File: as2/server/player.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+`%xt%sm%0%123%Hello%`
+
+
+### Send Safe Message
+Sent to all penguins in a room to show a safe message sent by a penguin. `ss`
+
+|Argument|Type|Variadic?|
+|---|---|---|
+|penguinID|Unsigned integer|No|
+|Safe Message ID|Unsigned integer|No|
```
Comment (widd) [inline as2/server/player.md]: Safe Message ID -> safeMessageID

File: as2/server/redemption.md [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+##### Messages pertaining to the redemption server.
+---
+### Join Server
+Sent to the penguin trying to log in to the redemption server to give needed player information. `rjs`
+
+|Argument|Type|Variadic?|Default
+|---|---|---|---
+|Redemeed Books|String|Yes|NULL
```
Comment (widd) [inline as2/server/redemption.md]: Redeemed Books -> redeemedBooks
(or perhaps a better short name? I am unable to think of one)

The argument type should also better describe the actual string - ie Unsigned integer array delimited by ","

File: as2/server/redemption.md [hunk 15]
```diff
@@ -0,0 +1,15 @@
+# Redemption
+##### Messages pertaining to the redemption server.
+---
+### Join Server
+Sent to the penguin trying to log in to the redemption server to give needed player information. `rjs`
+
+|Argument|Type|Variadic?|Default
```
Comment (widd) [inline as2/server/redemption.md]: A "default" column, while somewhat useful, is unncessary as there are no defaults to message arguments in a strict implementation. Should be removed for now.

File: as2/server/redemption.md [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+|---|---|---|---
+|Redemeed Books|String|Yes|NULL
+|Unknown & Not Needed|Unsigned integer|No|0
+|Member?|Unsigned integer|No|0
```
Comment (widd) [inline as2/server/redemption.md]: You can just change this to maybe "Boolean integer (0, 1)" and most people will understand without the need for the note. But it is good to be explicit.

**Review by widd — CHANGES_REQUESTED** [review]
Aside from the specific review comments I left, make sure to add a newline between the message/packet ID description and the ID itself. I believe all the files included need this change.

**Comment by jadbalout** [comment]
Done, anything else?

File: as2/server/player.md [hunk 17]
```diff
@@ -0,0 +1,98 @@
+# Player
+##### Messages pertaining to the player actions.
```
Comment (widd) [inline as2/server/player.md]: Should be "pertaining to actions performed by penguin avatars" (or something similiar) as that is much more clear.

File: as2/server/player.md [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+|Argument|Type|Variadic?|
+|---|---|---|
+|penguinID|Unsigned integer|No|
+|frame|Unsigned integer|No|
+
+##### Example
+`%xt%sf%0%123%5%`
+
+### Send Move
+Sent to all penguins in a room to update a player location.
```
Comment (widd) [inline as2/server/player.md:18]: Maybe add a note in the description that the client will animate waddling from point A to B on its own accord.

File: as2/server/redemption.md [hunk 19]
```diff
@@ -0,0 +1,16 @@
+# Redemption
+##### Messages pertaining to the redemption server.
+---
+### Join Server
+Sent to the penguin trying to log in to the redemption server to give needed player information. 
+
+`rjs`
+
+|Argument|Type|Variadic?
```
Comment (widd) [inline as2/server/redemption.md:9]: It seems this table is missing some formatting characters (the ending pipe or | characters).

File: as2/server/redemption.md [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+|---|---|---|---
+|redemeedBooks|Unsigned integer array delimited by ","|Yes
+|Unknown & Not Needed|Unsigned integer|No
+|member?|Boolean integer (0, 1)|No
```
Comment (widd) [inline as2/server/redemption.md]: `memberStatus` or `member` would be better than `member?` in this case.

**Review by widd — CHANGES_REQUESTED** [review]

**Comment by jadbalout** [comment]
done again

**Pull request merged**